### PR TITLE
BUG: fix skiprows callable infinite loop

### LINKED ISSUE #45585: BUG: read_excel skiprows callable can result in infinite loop
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import tempfile

import pandas as pd
from pandas.testing import assert_frame_equal

source_df = pd.DataFrame([{"row": 0}, {"row": 1}, {"row": 2}, {"row": 3}])

with tempfile.NamedTemporaryFile("w", suffix=".xlsx") as xlsx_fd:
    source_df.to_excel(xlsx_fd.name, index=False)

    df = pd.read_excel(
        xlsx_fd.name, skiprows=lambda x: x not in [0, 2]
    )  # infinite loop here

    expected_df = pd.DataFrame([{"row": 1}, {"row": 3}])

    assert_frame_equal(df, expected_df)
```

### Issue Description

In certain cases the `skiprows` callable on `read_excel` will loop infinitely. It's related to the sequence of `True` and `False` values coming out of the function. I've already written a unit test that triggers the issue and a fix to address it. Creating this bug to track the fix against.

### Expected Behavior

The test provided above should pass, where currently it will loop forever.

To put it another way; `skiprows` callable should never be called with a row index that is greater than the length of the document.

### Installed Versions

<details>
❯ ls bug.py | entr -c python bug.py
/Users/bram/Code/nca/pandas/pandas/compat/_optional.py:149: UserWarning: Pandas requires version '1.3.1' or newer of 'bottleneck' (version '1.2.1' currently installed).
  warnings.warn(msg, UserWarning)

INSTALLED VERSIONS
------------------
commit           : 04f5721617a622912f493f42f160139e073d0c2f
python           : 3.9.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.2.0
Version          : Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.5.0.dev0+130.g04f5721617
numpy            : 1.22.1
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.1.0
Cython           : 0.29.26
pytest           : 6.2.5
hypothesis       : 6.36.0
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : 1.2.1
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
zstandard        : None
</details>

### PR DESCRIPTION
**Pull request by bram2000**
- [X] closes #45585
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Not sure what to do about the `whatsnew` entry. Please advise (I've read the contributing doc but could not find it mentioned).

### CODE REVIEW

**Comment by bram2000** [comment]
I'm not sure why that "Test experimental data manager" test fails; it doesn't appear to be related to my code.

File: pandas/tests/io/excel/test_readers.py [hunk 1]
```diff
@@ -1148,6 +1148,22 @@ def test_read_excel_skiprows(self, request, read_ext):
         )
         tm.assert_frame_equal(actual, expected)
 
+        actual = pd.read_excel(
```
Comment (phofl) [inline pandas/tests/io/excel/test_readers.py:1151]: Please create a dedicated test

**Review by phofl — CHANGES_REQUESTED** [review]
Small comments 

please add a whatsnew

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (bram2000) [reply pandas/tests/io/excel/test_readers.py]: I've split this to a new test.

Comment (bram2000) [reply pandas/tests/io/excel/test_readers.py]: Wondering if the other "callable" test should be in there too, and we can call the test `test_read_excel_skiprows_callable`

**Comment by bram2000** [comment]
> Small comments
> 
> please add a whatsnew

Is the `whatsnew` process documented anywhere? I couldn't figure out what to do!

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -194,6 +194,7 @@ Conversion
 - Bug in constructing a :class:`Series` from a float-containing list or a floating-dtype ndarray-like (e.g. ``dask.Array``) and an integer dtype raising instead of casting like we would with an ``np.ndarray`` (:issue:`40110`)
 - Bug in :meth:`Float64Index.astype` to unsigned integer dtype incorrectly casting to ``np.int64`` dtype (:issue:`45309`)
 - Bug in :meth:`Series.astype` and :meth:`DataFrame.astype` from floating dtype to unsigned integer dtype failing to raise in the presence of negative values (:issue:`45151`)
+- Bug in :meth:`python_parser._next_line` causing an infinite loop with certain `skiprows` callables (:issue:`45585`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:197]: pls make this user facing, e.g. use `read_csv` / `read_excel` with `engine='python'`

File: pandas/tests/io/excel/test_readers.py [hunk 3]
```diff
@@ -1165,6 +1165,24 @@ def test_read_excel_skiprows(self, request, read_ext):
...
+    def test_read_excel_skiprows_callable_not_in(self, request, read_ext):
+        actual = pd.read_excel(
+            "testskiprows" + read_ext,
+            sheet_name="skiprows_list",
+            skiprows=lambda x: x not in [1, 3, 5],
+        )
+        expected = DataFrame(
+            [
+                [1, 2.5, pd.Timestamp("2015-01-01"), True],
+                # [2, 3.5, pd.Timestamp("2015-01-02"), False],
```
Comment (jreback) [inline pandas/tests/io/excel/test_readers.py:1177]: why are these commented?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.5.0.rst (hunk 2 above)
Comment (bram2000) [reply doc/source/whatsnew/v1.5.0.rst]: done

**Review by rhshadrach — APPROVED** [review]
lgtm

File: doc/source/whatsnew/v1.5.0.rst [hunk 4]
```diff
@@ -194,6 +194,7 @@ Conversion
 - Bug in constructing a :class:`Series` from a float-containing list or a floating-dtype ndarray-like (e.g. ``dask.Array``) and an integer dtype raising instead of casting like we would with an ``np.ndarray`` (:issue:`40110`)
 - Bug in :meth:`Float64Index.astype` to unsigned integer dtype incorrectly casting to ``np.int64`` dtype (:issue:`45309`)
 - Bug in :meth:`Series.astype` and :meth:`DataFrame.astype` from floating dtype to unsigned integer dtype failing to raise in the presence of negative values (:issue:`45151`)
+- Bug in :meth:`read_excel` with `engine=python` results in an infinite loop with certain `skiprows` callables (:issue:`45585`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.5.0.rst:197]: Can you move this to the IO section down below

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.5.0.rst [hunk 5]
```diff
@@ -227,6 +227,7 @@ MultiIndex
 I/O
 ^^^
 - Bug in :meth:`DataFrame.to_stata` where no error is raised if the :class:`DataFrame` contains ``-np.inf`` (:issue:`45350`)
+- Bug in :meth:`read_excel` with `engine=python` results in an infinite loop with certain `skiprows` callables (:issue:`45585`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.5.0.rst:230]: What does engine=python mean? I do not think that python is valid. Please also add ticks around the string, e.g. ``engine="python"``

Comment (bram2000) [reply doc/source/whatsnew/v1.5.0.rst]: I'm not sure it was suggested here https://github.com/pandas-dev/pandas/pull/45586#discussion_r790897640

I've just removed it.

Comment (phofl) [reply doc/source/whatsnew/v1.5.0.rst]: Please let the reviewer resolve conversations

I think I get what jeff meant. ``read_csv with engine="python" and read_excel``. 

Please also add a test for read_csv

**Comment by jreback** [comment]
@phofl @rhshadrach ok here?

**Comment by phofl** [comment]
Tests are failing.

- adjustment of whatsnew
- adding a test for read csv

are open points

**Comment by bram2000** [comment]
The merge of main seems to have broken something, I can no longer run my test...

```python -m pytest pandas/tests/io/excel/test_readers.py -vv -s
ImportError while loading conftest '/Users/bram/Code/nca/pandas/pandas/conftest.py'.
pandas/conftest.py:608: in <module>
    params=[
pandas/conftest.py:611: in <listcomp>
    if not isinstance(indices_dict[key], MultiIndex) and indices_dict[key].is_unique
pandas/_libs/properties.pyx:37: in pandas._libs.properties.CachedProperty.__get__
    val = self.fget(obj)
pandas/core/indexes/base.py:2297: in is_unique
    return self._engine.is_unique
pandas/_libs/properties.pyx:37: in pandas._libs.properties.CachedProperty.__get__
    val = self.fget(obj)
pandas/core/indexes/base.py:873: in _engine
    return libindex.ExtensionEngine(target_values)
E   AttributeError: module 'pandas._libs.index' has no attribute 'ExtensionEngine'```

**Comment by phofl** [comment]
You have to build pandas again
``
python setup.py build_ext --inplace -j 4
``

**Comment by bram2000** [comment]
I've added a similar test for csv... it passes but it doesn't appear to go through my new code anyway. Does it still make sense to mention `read_csv` in the whatsnew?

**Comment by phofl** [comment]
No, don't have to mention in this case

pre-commit is failing

**Comment by bram2000** [comment]
precommit is failing with the following, but I'm not sure why as I've copied the form of the other notes in the file.

```
rst ``code`` is two backticks...................................................................................Failed
- hook id: rst-backticks
- exit code: 1

doc/source/whatsnew/v1.5.0.rst:287:- Bug in :meth:`read_excel` results in an infinite loop with certain `skiprows` callables (:issue:`45585`)
```

**Comment by phofl** [comment]
:func:`read_excel` please

pre-commit is failing because of
`skiprows` -> ``skiprows``

**Comment by rhshadrach** [comment]
> I've added a similar test for csv... it passes but it doesn't appear to go through my new code anyway. Does it still make sense to mention `read_csv` in the whatsnew?

You have to add `engine="python"`. When I do this, I'm seeing `_next_line` being hit.

@phofl 

> Edit: Two backticks before and after skiprows

I struggled with this before. In GitHub (and maybe elsewhere?), you can use two different number of backticks to make backticks show up, e.g. `` `backtick` ``

**Comment by jreback** [comment]
@bram2000 can you merge master and address comments

**Comment by bram2000** [comment]
I've merged master.

Even if i manually add `engine="python"` to the `read_excel` call it still doesn't hit my debugger in the affected `while` loop. The test iterates through the four engine types anyway, so don't think it's necessary to explicitly set `engine`.

**Comment by rhshadrach** [comment]
> it still doesn't hit my debugger in the affected while loop

Ah, I see - yes, I'm seeing the same. However doesn't the same bug exist in L689 down below?

**Comment by bram2000** [comment]
> > it still doesn't hit my debugger in the affected while loop
> 
> Ah, I see - yes, I'm seeing the same. However doesn't the same bug exist in L691 down below?

No it doesn't seem to. I've tried various combinations of skiprow functions and I can't get it to fail.

**Comment by rhshadrach** [comment]
@bram2000 - thanks for checking! Sorry for my confusion, it all makes sense now.

**Review by rhshadrach — APPROVED** [review]

**Comment by rhshadrach** [comment]
@jreback @phofl - good here?

**Comment by rhshadrach** [comment]
@bram2000 - failure look unrelated. Can you merge main.

**Review by jreback — APPROVED** [review]

**Review by phofl — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @bram2000

**Pull request merged**